### ISSUE TITLE
Automatic installation of server

### ISSUE THREAD

**Issue by razzmatazz** [OmniSharp/omnisharp-emacs#275]
At least VS Code and Atom do this.

The code isn't pretty but its here https://github.com/OmniSharp/omnisharp-node-client/blob/0fa4cfbff6782420057ef731fbf4311fc376061e/lib/helpers/runtime.ts

**Comment by sp3ctum** [comment]
Definitely :+1: 
Installation has always been a pain in the ass.

**Comment by sp3ctum** [comment]
Just a thought: we basically do some kind of installation already, I mean for our travis build. I wonder if we could port that logic to elisp, and then the travis build could work basically the same way. But then we'd have the benefit of having an automatic installation as well.

And it would continuously get tested too, which is a huge thing!

**Comment by razzmatazz** [comment]
Partially implemented, works on macOS (for me). Should be OK for Linux too. Needs little work on Windows.

Didn't apear to be THAT hard to implement as I feared, actually.

**Comment by razzmatazz** [comment]
And yes, we should do the same thing on travis build. Definitely next on the list (with Windows implementation)

**Comment by razzmatazz** [comment]
Actually extracting a zip on emacs/windows installation appears to be problematic, unless you expect the user to download and extract unzip.exe to a known location... I'd wish to implement something that would work automatically, w/o any involvement of a user, even on Windows....
 - http://lists.gnu.org/archive/html/emacs-devel/2017-07/msg00292.html

**Comment by razzmatazz** [comment]
really don't want to write pkzip parser in elisp, unless thats the last resort..

**Comment by sp3ctum** [comment]
Would it work if the installation first downloaded unzip and then the
server package?

**Comment by razzmatazz** [comment]
@sp3ctum yes, I think it would, but I was a little bit worried that:
 - I couldn't find naked unzip.exe on the net over HTTPS (yet);
   - some sites allow to download this, but only inside a .zip, which kind induces of a chicken-n-egg problem for us :)
 - downloading an executable over the internet this injects untrusted executable from untrusted site into user's machine..
   - granted, we download omnisharp.exe inside omnisharp-roslyn-x86.zip but this is the part we can somewhat trus..

In other words, I am slight worried because of security implications..

**Comment by razzmatazz** [comment]
I think we could: find some unzip.exe, maybe from gnu win32 tools and host it on github releases page for our site.. this would be over https and I *think* GPL allows for downloading the binaries over the net and using those w/o doing any kind of linking with the app (emacs).. and emacs is in GPL, so .. don't know..

**Comment by kimgronqvist** [comment]
Windows has a decent shell in Powershell (which is included by default). If it would be acceptable to only support Powershell version 5 and up, you could simply run `powershell -command "Expand-Archive c:\a.zip -DestinationPath c:\a"`.

The powershell version can be checked by running:

```
(substring 
    (shell-command-to-string "powershell -command \"(Get-Host).Version.Major\"")
   0 -1)
```

If the the user has a version below the threshold, simply display a message that they need to extract the zip file themselves.

There are other ways to extract zip files in powershell using .NET, but checking the installed .NET version could be more trouble than it's worth.

**Comment by razzmatazz** [comment]
@kimgronqvist thanks! I think this will be enough for me, also I found a way not to require powershell v5, by using powershell script that references system.io.compression.fs:
 - https://stackoverflow.com/a/27768628/535154

will try to implement this now

**Comment by kimgronqvist** [comment]
Nice! I think it would also be worthwhile to run `(executable-find "unzip")` first, since this will be installed on some Windows systems that use Emacs.

Some users will have unzip in their path when installing Git for Windows (which includes msys2), and some will have installed msys2 manually since Emacs depends on tools like find, grep, and diff.

**Comment by razzmatazz** [comment]
@kimgronqvist another good idea!

**Comment by sp3ctum** [comment]
Great stuff, guys! I'm really glad you brought Powershell up Kim.
